tests/api/auth/test_basic.py

```diff
@@ -0,0 +1,162 @@
+"""Tests for basic authentication for the Ralph API."""
+
+import base64
+import json
+
+import bcrypt
+import pytest
+
+from ralph.api.auth.basic import ServerUsersCredentials, UserCredentials
+from ralph.conf import settings
+
+STORED_CREDENTIALS = json.dumps(
+    [
+        {
+            "username": "ralph",
+            "hash": bcrypt.hashpw(b"admin", bcrypt.gensalt()).decode("UTF-8"),
+            "scopes": ["ralph_test_scope"],
+        }
+    ]
+)
+
+
+def test_api_auth_basic_model_serveruserscredentials():
+    """Test api.auth ServerUsersCredentials model."""
+
+    users = ServerUsersCredentials(
+        __root__=[
+            UserCredentials(
+                username="johndoe", hash="notrealhash", scopes=["johndoe_scope"]
+            ),
+            UserCredentials(username="foo", hash="notsorealhash", scopes=["foo_scope"]),
+        ]
+    )
+    other_users = ServerUsersCredentials.parse_obj(
+        [
+            UserCredentials(
+                username="janedoe", hash="notreallyrealhash", scopes=["janedoe_scope"]
+            ),
+        ]
+    )
+
+    # Test addition operator
+    users += other_users
+
+    # Test len
+    assert len(users) == 3
+
+    # Test getitem
+    assert users[0].username == "johndoe"
+    assert users[1].username == "foo"
+    assert users[2].username == "janedoe"
+
+    # Test iterator
+    usernames = [user.username for user in users]
+    assert len(usernames) == 3
+    assert usernames == ["johndoe", "foo", "janedoe"]
+
+    # Test username uniqueness validator
+    with pytest.raises(
+        ValueError,
+        match="You cannot create multiple credentials with the same username",
+    ):
+        users += ServerUsersCredentials.parse_obj(
+            [
+                UserCredentials(
+                    username="foo", hash="notsorealhash", scopes=["foo_scope"]
+                ),
+            ]
+        )
+
+
+@pytest.mark.parametrize("auth_mode", ["basic"])
+def test_get_whoami_no_credentials(auth_test_client):
+    """Whoami route returns a 401 error when no credentials are sent."""
```

Comment:
maybe we could have a `basic_auth_test_client` fixture instead of parametrizing all tests ? (I am not sure this use is in the spirit of `parametrize` but I might be wrong).

Reply:
I agree it's not in the spirit of `parametrize`, re-splitting into two distinct fixtures as I find it more clear.